### ISSUE TITLE
Hub SDK: story.yaml + story.lock

### ISSUE THREAD

**Issue by wilzbach** [storyscript/hub-sdk-python#27]
tl;dr:
- provide a mechanism for the CLI/Compiler and SLS to interact with the SDK and trigger a download
- the current approach to download the world of packages doesn't scale well if we consider that every package might have n different versions
- services need to be auto-locked in the story directory to ensure predictable execution
- calls to the service hub should be minimized
- new services that appear in a story should be auto-added to the story package list + result in a regeneration of the lock file

story.yaml could look roughly look:

```yaml
packages:
   - twitter: 2.0.7
   - redis:
       min: 3.0.0
       max: 4.0.0
  - postgres: latest
```

TODO:
- [ ] do we use shorthand SemVer range specifiers or more explicit keys

and `story.lock` should contain at least
- md5 hash of `story.yaml`
- `version` format
- list of all locked packages with their full information for a respective `version`

```json
{
   "version": "1",
   "packages": {
       "twitter": {
           "version": "2.0.8",
            "title": "Twitter API",
           "description": "...",
           "commands": {"...": "..."},
            "...": "..."
       }
   }
}
```

We need to expose at least these functions for the compiler and SLS:

### `check_packages([<package>])`

- calls `check_lock_file(packages)`
- returns the deserialized packages

#### `check_lock_file(packages)`

- open `story.lock`
  -  check md5 hash of `story.yaml`. if mismatch -> jump to `generate_lock_file`
  - if any of the packages doesn't occur in `story.lock`
     - for all of these packages -> `add_to_story_yaml`
     -> jump to `generate_lock_file`

-----

And this internal functionality:

#### `generate_lock_file`

- open `story.yaml` -> read packages
- graphQL query to hub for all versions of each package
- calculate best-matching version for each package based on SemVer
- save as `story.lock` with `version` and `md5` of `story.yaml`

#### `add_to_story_yaml(packages)`

- graphQL query to hub for latest versions of all package
- add these packages to the `story.yaml`
